# Задание
Провести исследование на собранном ранее наборе данных (он может быть дополнен или собран новый)

- определить цель исследования
- разделить набор данных на тренировочный и тестовый 
- использовать на выбор любой метод линейной регрессии 
- описать ход исследования дав пояснение и обоснование
- сделать выводы и определить практическое значение полученных знаний

цель: 
    определить, как сделать пиццу, чтобы за 3500 рублей стоила своих денег.

In [2]:
import pandas as pd

piz_df=pd.read_csv('https://raw.githubusercontent.com/GVicSt/DataSets/main/pizas_data_small.csv', sep='\t')
piz_df

имя  размер(см)  популярность  эксклюзивность  цена(руб)  \
0   Margherita                 30.5           0.8             0.2     749.70   
1   Pepperoni                  35.2           0.9             0.3     936.48   
2   Hawaiian                   28.0           0.7             0.1     819.89   
3   Veggie Supreme             32.8           0.6             0.2     878.71   
4   BBQ Chicken                33.5           0.7             0.3     982.88   
5   Meat Lovers                36.0           0.8             0.4    1124.01   
6   Four Cheese                29.7           0.6             0.1     787.51   
7   Pesto Delight              31.2           0.5             0.2     852.21   
8   Supreme                    34.0           0.8             0.3    1055.12   
9   Mediterranean              30.8           0.7             0.2     868.61   
10  Veggie Delight             32.5           0.6             0.1     846.81   
11  Capricciosa                33.2           0.7             0.3     965.92   
12  Quattro Formaggi           28.5           0.6             0.2     810.61   
13  Spicy Sausage              31.0           0.8             0.3    1001.17   
14  Bianca                     34.5           0.7             0.2    1040.39   

    калории  масса(грамм)  
0       250           350  
1       320           420  
2       280           380  
3       290           400  
4       350           460  
5       420           550  
6       280           390  
7       310           420  
8       380           500  
9       300           410  
10      290           400  
11      360           480  
12      320           430  
13      370           490  
14      340           450

In [21]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

def get_regression_output(main_index, other_indexes, data, split_p, input):
    # разделяю данные на целевые и другие
    x = data[main_index]
    y = data[other_indexes].values.reshape(-1,1)
    # разделяю данные на тренировочные и тестовые
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = split_p, random_state = 0)

    # использую линейную регрессию
    regressor = LinearRegression()
    regressor.fit(x_train, y_train)

    # получаю результаты
    output =      pd.DataFrame(regressor.predict(pd.DataFrame(input)))
    output_test = pd.DataFrame(regressor.predict(x_test))

    return [output.T,output_test,y_test]

new_pizza = pd.DataFrame({'размер(см)':[75.0],'популярность':[1.0],'эксклюзивность':[1.0],'масса(грамм)':[1187.0],'калории':[895.0]})

out = get_regression_output(['размер(см)','популярность','эксклюзивность','масса(грамм)','калории'],
                            'цена(руб)',
                            piz_df,
                            0.2,
                            new_pizza
                            )

print(f'цена которая нужна для соответствия качество/цена, при характеристиках:\n\n{new_pizza}\n\nравна: {out[0][0].values[0]}\n\n\
      \nданные теста:\n  предсказанные данные:\n{out[1].values}\n\n  исходные данные:\n{out[2]}')

цена которая нужна для соответствия качество/цена, при характеристиках:

   размер(см)  популярность  эксклюзивность  масса(грамм)  калории
0        75.0           1.0             1.0        1187.0    895.0

равна: 2395.10576196951

      
данные теста:
  предсказанные данные:
[[1023.19641282]
 [ 794.06083815]
 [1063.93370229]]

  исходные данные:
[[ 936.48]
 [ 787.51]
 [1055.12]]


я так полагаю что модель в основном во всём похожа, кроме первой строки, и то погрешность всего ~100 рублей.

то есть при условии, что я увеличил размер массу и каллорийность пиццы(которая стоит ~1001) в ~2.4 раза,
цена должна быть также в ~2.4 раз больше 1001, а именно 2422.
итого 2395 от 2422 отличается всего на 27 рублей (~1.12%).

и результат по цели вполне меня устраивает.

In [35]:
import math

print(f'пицца больше в {(2*math.pi*75)/(2*math.pi*31)} раза, относитьльно пиццы за 1000')

# print(490/(2*math.pi*31)) # = 2.52

# print((2*math.pi*31)*2.52) # = 490
print(f'масса должна быть {(2*math.pi*75)*2.52}')

# print(370/(2*math.pi*31))  # = 1.9

# print((2*math.pi*31)*1.9)  # = 370
print(f'каллорий должно быть {(2*math.pi*75)*1.9}')

print(f'цена по расчётам {(1001.17*2.4193548387096775)}')

print(f'отклонение предсказания от расчётов {(1-(2395.10576196951/(1001.17*2.4193548387096775)))*100}%')

пицца больше в 2.4193548387096775 раза, относитьльно пиццы за 1000
масса должна быть 1187.5220230569419
каллорий должно быть 895.353906273091
цена по расчётам 2422.185483870968
отклонение предсказания от расчётов 1.1179871269883512%
